### ISSUE TITLE
Add linux-ec2 and linux-ec2-lts kernels

### ISSUE THREAD

**Issue by AdamSLevy** [archzfs/archzfs#268]
I use Arch Linux on AWS ec2 and use ZFS on many of my servers using the pre-built zfs-dkms packages distributed from your repos. 

How would I go about creating my own zfs module package built against the linux-ec2{-lts} kernels distributed from Uplink Labs: 
https://www.uplinklabs.net/projects/arch-linux-on-ec2/
https://git.uplinklabs.net/steven/projects/archlinux/ec2/ec2-packages.git/

If you can point me in the right direction at first and guide me a bit on the process, I would be happy to contribute back to the project by documenting the steps of adding a new kernel for the wiki.

Thank you, I really appreciate this project.

**Comment by minextu** [comment]
First you need to setup archzfs to be able to build packages (https://github.com/archzfs/archzfs/wiki/Building-archzfs). You will need to compile your custom kernel using clean-chroot-manager so it will be able to find it later.

To create a new kernel config just copy and rename an existing kernel in src/kernels (e.g. linux-lts.sh). You just need to make a few adjustments like updating the url + regex to get the latest kernel version, set the correct kernel mod path and change names.

You can then generate the PKGBUILD with ./build.sh -d update $mode_name (-d will show more info) 
The pkgbuilds will be generated to $zfs_pkgbuild_path. 
Finally create the package using ./build.sh - d make $mode_name

If something isn't working or you have more question, feel free to ask them here!

**Comment by minextu** [comment]
You can add existing packages in the current directory using the command `ccm64 a`

**Comment by AdamSLevy** [comment]
Okay thank you. So I set up clean-chroot-manager and edited its config and the conf.sh file according to the instructions in the wiki. 

I created a new script: `src/kernels/linux-ec2-lts.sh` which I initially copied from the `src/kernels/linux-lts.sh` script.

I searched and replaced all instances of `linux-lts` with `linux-ec2-lts`. I hardcoded the version rather than trying to parse it from a webpage.

I ran the update and make steps for common and common-git.

I downloaded the latest linux-ec2-lts kernel that I want to build against, and added it to the package repo for the chroot using `ccm64 a` and the output verified that the package was added.

Then I ran `sudo ./build.sh -d ec2-lts update` and ran into the following error:

```
==== ERROR: Failure while fetching https://github.com/zfsonlinux/zfs.git repo
```

Full output can be found here: https://gist.github.com/AdamSLevy/7ae41f164b5b19fdbec3e488cca7b783

It seems like this might have something to do with it trying to build the -git versions of these packages, which I am not interested in doing.

Your help is greatly appreciated.

**Comment by AdamSLevy** [comment]
To reduce confusion. I forked the repo and have pushed the changes I made. You can see them here: https://github.com/AdamSLevy/archzfs/tree/ec2-lts

Thank you again

**Comment by minextu** [comment]
Seems like the relevant error is 
```
cp: cannot create directory '/home/aslevy/repos/canonical-ledgers/archzfs/packages/linux-ec2-lts/zfs-linux-ec2-lts-git/zfs': No such file or directory
```
So you either need to create the zfs-ec2-lts-git folder or remove / comment out the `update_linux_lts_git_pkgbuilds` function in linux-ec2-lts.sh to disable git builds for this kernel. Unfortunately there is no command line option for this right now.

**Comment by AdamSLevy** [comment]
What normally creates those -git directories? The directories `packages/linux-ec2-lts/zfs-linux-ec2-lts` and `packages/linux-ec2-lts/spl-linux-ec2-lts` already exist. Why don't the -git directories.

Also the -git packages are building the latest -git version of the zfs and spl modules, is that right? As opposed to building against a -git kernel version...

**Comment by AdamSLevy** [comment]
Success! Thank you so much for your help and for maintaining ZFS on Arch Linux!

**Comment by minextu** [comment]
Happy to hear that you got it working. You can open a pull request with the new kernel config, if you want.

**Comment by AdamSLevy** [comment]
I would need to clean it up a bit and implement automatically getting the latest version number. 

If I did that and it was merged here, would this configuration start being packaged and be downloadable from the archzfs package repos?

That would certainly make it worth it to finish cleaning it up. For my purposes putting in the latest kernel version manually is alright for me.

**Comment by minextu** [comment]
Yep it would be available in the archzfs repo. But probably only in the automated archzfs-testing repo, since I don't know if I can test the resulting packages without having an ec2 instance.

**Comment by minextu** [comment]
I just pushed https://github.com/archzfs/archzfs/commit/19777aec8be56da198d366f9f0f506e8f3bf49f8, which might help you. I needed to get the kernel version of linux-vfio using a repo. So you can now use something similar to this:
```
   msg "Checking linux-vfio repo for the latest linux kernel version..."
    if ! get_repo_package "https://repo.markzz.com/arch/markzz/x86_64/markzz.files" "(?<=linux-vfio-)[\d\w\.-]+(?=/)"; then
        exit 1
    fi
kernel_version=${regex_match}
```

Just replace the repo url with the one for ec2 (also replace linux-vfio with linux-ec2) and it should just work.

**Comment by AdamSLevy** [comment]
Thank you, that's terrific! I'll get a PR opened this week.

**Comment by AdamSLevy** [comment]
So there may be a small issue with you adding these kernels to the package repo. It appears that the S3 bucket hosting the ec2 packages has permissions set such that downloads can only be made from a valid ec2 ip address. From other IP addresses I get access denied when I try to download packages.

**Comment by minextu** [comment]
Ah bummer... It won't work that easily then

**Comment by AdamSLevy** [comment]
I'm having trouble getting this to work a second time with the latest commits on master. 

I rebased my changed and then added a function to parse out the kernel version.

It won't successfully output any PKGBUILDs.

Here is my output from running the command

https://gist.github.com/4f50a11c9ee1bc343afe1e575ff56f71

**Comment by AdamSLevy** [comment]
Thank you again!

**Comment by AdamSLevy** [comment]
So I discovered that the ec2-lts kernel can be downloaded, just not using a simple curl command. There must be some sort of required headers that pacman provides that allows the download to happen from this repo:

```
[ec2]
SigLevel = PackageRequired
Server = https://arch-linux-ami.s3.amazonaws.com/repo/$arch
```

If there is still interest in merging in this kernel I will submit a PR with my commits for this kernel and could add the non-lts ec2 kernel as well.

**Comment by minextu** [comment]
Sure, but I will probably only add it to the testing repository, since I don't have an easy way to test the resulting builds.

**Comment by AdamSLevy** [comment]
@minextu I am having an issue building my ec2-lts zfs packages now that SPL has been removed.

I have rebased my changes on master yet it persists in trying to build the SPL packages, for which it currently fails. I'm trying to solve my own issue by examining the changes made on master but if you have a moment to take a look and make a suggestion it would be greatly appreciated.

Thank you

**Comment by minextu** [comment]
build.sh shouldn't try to build spl at all except if it appears in `pkg_list` in your kernel config. I can take a look at your code if you want, just sent me a link 🙂

**Comment by AdamSLevy** [comment]
I solved the most recently mentioned issue regarding build spl packages. I have been successfully building ec2-lts packages for a bit now. I've uploaded my latest changes here: https://github.com/AdamSLevy/archzfs.git on the `ec2-lts` branch.

Previously I reported that the linux-ec2-lts package wasn't available to download outside of an ec2 instance. This was incorrect. It is possible to automate the download of the linux package, however I don't know enough about your build system to know the right way to add that.

The ec2 repo file that lists the latest linux-ec2 and linux-ec2-lts package can be found here: https://arch-linux-ami.s3.amazonaws.com/repo/x86_64/ec2.db
 and the package is of course next to it. 

Currently I have a small hacky script that pulls the latest kernel version number for ec2-lts using the git repo found here: https://git.uplinklabs.net/steven/projects/archlinux/ec2/ec2-packages.git
 but this is not the best way to check the kernel version, as this git repo is sometimes updated slightly ahead of the actual package repo.

Happy to discuss how to help the integration of linux-ec2 and linux-ec2-lts packages if this is something you are still interested in.
